codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/utils/ModelUtils.java

```diff
@@ -326,12 +340,16 @@ public static Set<List<ShapeId>> findAllDependentShapesWithPaths(
             .map(Collections::singletonList)
             .collect(Collectors.toSet());
         Set<List<ShapeId>> pathsToShapes = new LinkedHashSet<>(new LinkedHashSet<>());
+        Set<List<ShapeId>> pathsSummary = new LinkedHashSet<>(new LinkedHashSet<>());
 
         // Breadth-first search via getDependencyShapeIds
         final Queue<List<ShapeId>> toTraverse = new LinkedList<>(initialShapeIdsAsPaths);
         while (!toTraverse.isEmpty()) {
             final List<ShapeId> currentShapeIdWithPath = toTraverse.remove();
-            if (pathsToShapes.add(currentShapeIdWithPath)) {
+
+	    // to avoid cycles, only keep the first list with a given first and last element
```

Comment:
I can guarantee that cycles are traversed, and you get an arbitrarily large set of arbitrarily large lists.

I'm pretty sure that the problem is with
dependencyShapeIdsWithPaths.forEach(toTraverse::add);
where given (A,B,C) for every dependent X of C we add the list (A,B,C,X) and so we add (A,B,C,B) and (A,B,C,B,C) et cetera.

Reply:
Ah got it, `toTraverse` is a set of paths rather than nodes, so it doesn't prevent traversing cycles.

Reply:
You can make this simpler and clearer by replacing `pathsSummary` by a `Set<ShapeId> visitedShapes` instead. All paths this method produces will always start from the same ShapeID so that's exactly equivalent, and to boot it matches the classic BFS algorithm so it'll be easier to understand.

Reply:
Done